validation-test/compiler_crashers_2_fixed/sr12723.swift

```diff
@@ -0,0 +1,83 @@
+// RUN: %target-swift-emit-silgen %s -verify
+
+func SR12723_thin(_: (@convention(thin) () -> Void) -> Void) {}
+func SR12723_block(_: (@convention(block) () -> Void) -> Void) {}
+func SR12723_c(_: (@convention(c) () -> Void) -> Void) {}
+
+func SR12723_function(_: () -> Void) {}
+
+func context() {
+    SR12723_c(SR12723_function) 
+
+    SR12723_block(SR12723_function) 
+
+    SR12723_thin(SR12723_function) 
+}
+
+struct SR12723_C {
+    let function: (@convention(c) () -> Void) -> Void
+}
+
+struct SR12723_Thin {
+    let function: (@convention(thin) () -> Void) -> Void
+}
+
+struct SR12723_Block {
+    let function: (@convention(block) () -> Void) -> Void
+}
+
+func proxy(_ f: (() -> Void) -> Void) {
+    let a = 1
+    f { print(a) }
+}
+
+func cContext() {
+    let c = SR12723_C { app in app() }
+
+    proxy(c.function)
+    // expected-error@-1 {{cannot convert value of type '(@convention(c) () -> Void) -> Void' to expected argument type '(() -> Void) -> Void'}}
+
+    let _ :  (@convention(block) () -> Void) -> Void = c.function 
+    // expected-error@-1 {{cannot convert value of type '(@convention(c) () -> Void) -> Void' to specified type '(@convention(block) () -> Void) -> Void'}}
+
+    let _ :  (@convention(c) () -> Void) -> Void = c.function // OK
+
+    let _ :  (@convention(thin) () -> Void) -> Void = c.function 
+    // expected-error@-1 {{cannot convert value of type '(@convention(c) () -> Void) -> Void' to specified type '(@convention(thin) () -> Void) -> Void'}}
+
+    let _ :  (() -> Void) -> Void = c.function 
```

Comment:
This and some other conversions here are `thin-to-thick` which should be allowed. I think this is actually meant to work in reverse - `thin-to-thick` is allowed but `thick-to-thin` isn't.

Reply:
At least this is what Joe mentioned in his comment.

Reply:
Hey @xedin :) 
Hum, that's actually one of the questions I had, because on the param function type there is the fact that subtype relation is contravariant. So my question is if this affects the conversion in some way? 
In this case, for example, converting `(@convention(c)()->Void)->Void` to `(()->Void)->Void` on `A -> B` to `A'-> B'` respectivelly, this means the subtype relation is `A'` is subtype of `A` so the `()->Void` is a subtype of `@convention(c)()->Void`, does this means the conversion is `thick-to-thin`? ... at the time this makes sense to me, but that is actually a question I have, do you know if this makes sense or am I missing something? Let me know if I this makes sense :) 

Also, the logic on `matchFunctionTypes` is only allowing exactly `thin-to-thick` and `thick-to-thick` any other conversions are being deferred to check if representations are the same.

Reply:
Yes, you are correct, sorry for the confusion. I was just merely trying to point out that we might want to change diagnostic or add a custom note which explains precisely why these types couldn't be converted (related to contravariance) because it appears that it's passing `thick-to-thin`.

Reply:
I got tripped by this myself because I didn't notice that diagnostic mentioned whole function type...

Reply:
Right, thanks for the clarification :)
I've added in the last commit, thin to thin which seems to be allowed, so now we strictly only validating `thick-to-thin`.

> we might want to change diagnostic or add a custom note which explains precisely why these types couldn't be converted (related to contravariance) because it appears that it's passing 

About the note, I think it makes sense, but I'm not sure how to phrase it because I'm not sure if we should mention variance, or only mention that the representations are not convertible in a parameter subtype context ...

Reply:
Yeah, it's a good question how note should be worded exactly...

Reply:
After thinking about this some more - I think we should just open a follow up SR for the note and merge what we have now.

Reply:
I think it makes sense :)